File: EverCam/IAP Helper/IAPController/IAPController.m
/*========================================
 
 - Instafun -
 
 made by FV iMAGINATION ©2015
 All rights reserved
 
 ========================================*/


#import "IAPController.h"
#import "Configs.h"
#import "HomeVC.h"


@interface IAPController ()
@end

@implementation IAPController

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

-(void)viewWillAppear:(BOOL)animated {
    NSLog(@"IAP MADE: %d", iapMade);
    
    if (!iapMade) {  _buyOutlet.hidden = false;
    } else {         _buyOutlet.hidden = true;   }
}


- (void)viewDidLoad {
    [super viewDidLoad];
    
    // ADD PROGRESS HUD ================
    hud = [[MBProgressHUD alloc] initWithView:self.view];
    [self.view addSubview:hud];
    [hud show:true];
    
    
    // Set Texts
    _titleLabel.text = NSLocalizedString(@"EverCam Premium", @"");
    [_buyOutlet setTitle:NSLocalizedString(@"Get Premium Now", @"") forState:UIControlStateNormal];
    [_restorePurchaseOutlet setTitle:NSLocalizedString(@"Restore Purchase", @"") forState:UIControlStateNormal];
    

    // Fetch available IAP products
    [self fetchAvailableProducts];
}





#pragma mark - BUY PREMIUM VERSION BUTTON
- (IBAction)buyButt:(id)sender {
    SKPayment *payment = [SKPayment paymentWithProduct: _product];
    [[SKPaymentQueue defaultQueue] addTransactionObserver:self];
    [[SKPaymentQueue defaultQueue] addPayment:payment];
    
    //[hud show:true];
}


#pragma mark - RESTORE PURCHASE BUTTON
- (IBAction)restorePurchaseButt:(id)sender {
    [[SKPaymentQueue defaultQueue] addTransactionObserver:self];
    [[SKPaymentQueue defaultQueue] restoreCompletedTransactions];
    [hud show:true];
}





#pragma mark - IAP METHODS ----------------------------------------------
-(void)fetchAvailableProducts {
    
    if([SKPaymentQueue canMakePayments]){
        NSLog(@"User can make payments");
        SKProductsRequest *productsRequest = [[SKProductsRequest alloc] initWithProductIdentifiers:[NSSet setWithObject:IAP_PRODUCT]];
        productsRequest.delegate = self;
        [productsRequest start];
        
    } else {
        UIAlertView *alert = [[UIAlertView alloc]initWithTitle: APP_NAME
        message:@"Purchases are diabled on this device!"
        delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil, nil];
        [alert show];
    }

}

-(void)productsRequest:(SKProductsRequest *)request didReceiveResponse:(SKProductsResponse *)response {
    
    NSUInteger count = response.products.count;
    SKProduct *validProduct = nil;
    if(count > 0){
        validProduct = [response.products objectAtIndex:0];
        NSLog(@"Products Available!");
        _product = validProduct;
        
        _descTxt.text = [NSString stringWithFormat:@"%@ \nfor just %@", _product.localizedDescription, _product.price];
        _descTxt.font = [UIFont fontWithName:@"Titillium-Semibold" size: 17];
        _descTxt.textColor = [UIColor whiteColor];
        _descTxt.textAlignment = NSTextAlignmentCenter;
        
        
        // Product id is not valid, this shouldn't be called unless that happens.
    } else if(!validProduct) { NSLog(@"No products available");  }
    
    // Hide HUD
    [hud hide:true];
}


-(void)purchaseMyProduct:(SKProduct *)product {
    SKPayment *payment = [SKPayment paymentWithProduct: product];
    [[SKPaymentQueue defaultQueue] addTransactionObserver:self];
    [[SKPaymentQueue defaultQueue] addPayment:payment];
}


- (void)paymentQueue:(SKPaymentQueue *)queue updatedTransactions:(NSArray *)transactions{
    
    for(SKPaymentTransaction *transaction in transactions){
        switch(transaction.transactionState) {
            case SKPaymentTransactionStatePurchasing:
                NSLog(@"Transaction state -> Purchasing");
                [hud show:true];
                break;
                
            case SKPaymentTransactionStatePurchased:
                [[SKPaymentQueue defaultQueue] finishTransaction:transaction];

                [self purchaseResults];
                
                NSLog(@"Transaction state -> Purchased");
                break;
                
            case SKPaymentTransactionStateRestored:
                [[SKPaymentQueue defaultQueue] finishTransaction:transaction];
                NSLog(@"Transaction state -> Restored");
                [self purchaseResults];
                
                break;
                
            case SKPaymentTransactionStateFailed:
                if(transaction.error.code == SKErrorPaymentCancelled){
                    NSLog(@"Transaction state -> Cancelled");
                    // The user cancelled the payment ;(
                    [hud hide:true];
                }
                [[SKPaymentQueue defaultQueue] finishTransaction:transaction];
                break;
            case SKPaymentTransactionStateDeferred:
                [hud hide:true];
                [[SKPaymentQueue defaultQueue] finishTransaction:transaction];
                break;
        }
    }
}


-(void)paymentQueueRestoreCompletedTransactionsFinished:(SKPaymentQueue *)queue {
    [self purchaseResults];
}


-(void)purchaseResults {
    // BOOL iapMade = TRUE, save and store it forever
    iapMade = true;
    [[NSUserDefaults standardUserDefaults] setBool:iapMade forKey:@"iapMade"];
    NSLog(@"%d", iapMade);
    
    UIAlertView *alert = [[UIAlertView alloc]initWithTitle: APP_NAME
    message: [NSString stringWithFormat:@"You've unlocked %@ Premium!", APP_NAME]
    delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil, nil];
    [alert show];
    
    [self dismissViewControllerAnimated:true completion:nil];
    [hud hide:true];
}






#pragma mark -  DISMISS BUTTON
- (IBAction)dimissButt:(id)sender {
    [self dismissViewControllerAnimated:true completion:nil];
}





- (void)didReceiveMemoryWarning{
    [super didReceiveMemoryWarning];
}

@end

File: Shuotes/IAPController.swift
/*------------------------------------------

- Shoutes -

©2015 FV iMAGINATION
All Rights reserved

--------------------------------------------*/


import UIKit
import StoreKit


// MARK: - CUSTOM PREVIEW CELL
class PreviewCell: UICollectionViewCell {
    /* Views */
    @IBOutlet weak var prevImage: UIImageView!
    
}



// MAKR: - IAP CONTROLLER CLASS
class IAPController: UIViewController,
SKProductsRequestDelegate,
SKPaymentTransactionObserver
{
    
    /* Views */
    @IBOutlet var iapDescriptionLabel: UILabel!

    @IBOutlet weak var buyOutlet: UIButton!
    @IBOutlet var restoreOutlet: UIButton!
    
    @IBOutlet weak var previewCollView: UICollectionView!
    
    
    /* Variables */
    var productsRequest = SKProductsRequest()
    var validProducts = []
    var paidOverlaysArray = [Int]()

    

    
    
    
override func prefersStatusBarHidden() -> Bool {
        return true
}
override func viewDidLoad() {
        super.viewDidLoad()
    
    view.showHUD(view)
    
    // Round buttons corners
    buyOutlet.layer.cornerRadius = 6
    restoreOutlet.layer.cornerRadius = 6
    
    if IAP_MADE {  buyOutlet.hidden = true }
    
    // Fetch available IAP Products
    fetchAvailableProducts()
    
    
    // Prepare the array of Paid Overlays
    let paidOverlaysQty = ALL_OVERLAYS - FREE_OVERLAYS
    for i in 0..<paidOverlaysQty { paidOverlaysArray.append(i + FREE_OVERLAYS) }
    previewCollView.reloadData()
}

    
    
    
// MARK: - COLLECTION VIEW DELEGATES 
func numberOfSectionsInCollectionView(collectionView: UICollectionView) -> Int {
    return 1
}
    
func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
    return paidOverlaysArray.count
}
    
func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
    let cell = collectionView.dequeueReusableCellWithReuseIdentifier("PreviewCell", forIndexPath: indexPath) as! PreviewCell
        
    cell.prevImage.image = UIImage(named: "\(paidOverlaysArray[indexPath.row]).png")
    
return cell
}
    
func collectionView(collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAtIndexPath indexPath: NSIndexPath) -> CGSize {
    return CGSizeMake(92, 92)
}
    

    
    
    

// MARK: - BUY OVERLAYS BUTTON
@IBAction func buyButt(sender: AnyObject) {
    purchaseMyProduct(validProducts[0] as! SKProduct)
    buyOutlet.enabled = false
    view.showHUD(view)
}
    
    
    
    
// MARK: - IAP METHODS -----------------------------------------------
func fetchAvailableProducts()  {
    let productIdentifiers = NSSet(objects:
            IAP_PRODUCT_IDENTIFIER
    )
        
    productsRequest = SKProductsRequest(productIdentifiers: productIdentifiers as! Set<String>)
    productsRequest.delegate = self
    productsRequest.start()
}
func productsRequest (request:SKProductsRequest, didReceiveResponse response:SKProductsResponse) {
        let count:Int = response.products.count
        
        // Finding IAP Products you've created in iTunes Connect
        if (count > 0) {
            validProducts = response.products
            
            // IAP Product
            let iapProduct: SKProduct = response.products[0] 
            iapDescriptionLabel.text = "\(iapProduct.localizedDescription)\nFor just \(iapProduct.price)"
            buyOutlet.hidden = false
            hudView.removeFromSuperview()
        }
}
func canMakePurchases() -> Bool {
    return SKPaymentQueue.canMakePayments()
}
func purchaseMyProduct(product: SKProduct) {
    if self.canMakePurchases() {
        let payment = SKPayment(product: product)
        SKPaymentQueue.defaultQueue().addTransactionObserver(self)
        SKPaymentQueue.defaultQueue().addPayment(payment)
            
    // IAP Purchases dsabled on the Device
    } else {
        let alert = UIAlertView(title: "Purchases are disabled in your device!",
        message: nil,
        delegate: nil,
        cancelButtonTitle: "OK")
        alert.show()
    }
    
}
func paymentQueue(queue: SKPaymentQueue, updatedTransactions transactions: [SKPaymentTransaction])    {
        for transaction:AnyObject in transactions {
            
            if let trans:SKPaymentTransaction = transaction as? SKPaymentTransaction {
                switch trans.transactionState {
                    
                case .Purchased:
                    SKPaymentQueue.defaultQueue().finishTransaction(transaction as! SKPaymentTransaction)
                    IAP_MADE = true
                    defaults.setBool(IAP_MADE, forKey: "IAP_MADE"); defaults.synchronize()
                    buyOutlet.hidden = true
                    hudView.removeFromSuperview()
                    break
                case .Failed:
                    SKPaymentQueue.defaultQueue().finishTransaction(transaction as! SKPaymentTransaction)
                    hudView.removeFromSuperview()
                    break
                case .Restored:
                    SKPaymentQueue.defaultQueue().finishTransaction(transaction as! SKPaymentTransaction)
                    hudView.removeFromSuperview()
                    break
                default: break  }
            }
        }
}
    
    
// RESTORE PURCHASE BUTTON
@IBAction func restorePurchaseButt(sender: AnyObject) {
    view.showHUD(view)
    SKPaymentQueue.defaultQueue().addTransactionObserver(self)
    SKPaymentQueue.defaultQueue().restoreCompletedTransactions()
}
    
func paymentQueueRestoreCompletedTransactionsFinished(queue: SKPaymentQueue) {
    IAP_MADE = true
    defaults.setBool(IAP_MADE, forKey: "IAP_MADE"); defaults.synchronize()
    buyOutlet.hidden = true
    hudView.removeFromSuperview()
}
    
    
    
// DISMISS BUTTON
@IBAction func dismissButt(sender: AnyObject) {
    dismissViewControllerAnimated(true, completion: nil)
}
    

    
    
override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
}
